src/geocel/detail/GeantVolumeInstanceMapper.cc

```diff
@@ -0,0 +1,244 @@
+//------------------------------- -*- C++ -*- -------------------------------//
+// Copyright Celeritas contributors: see top-level COPYRIGHT file for details
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file geocel/detail/GeantVolumeInstanceMapper.cc
+//---------------------------------------------------------------------------//
+#include "GeantVolumeInstanceMapper.hh"
+
+#include <string>
+#include <G4LogicalVolume.hh>
+#include <G4ReplicaNavigation.hh>
+#include <G4VPVParameterisation.hh>
+#include <G4VPhysicalVolume.hh>
+
+#include "corecel/Assert.hh"
+#include "corecel/cont/Range.hh"
+#include "corecel/io/Join.hh"
+#include "corecel/io/Logger.hh"
+#include "corecel/math/Algorithms.hh"
+#include "geocel/VolumeVisitor.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+namespace
+{
+//---------------------------------------------------------------------------//
+//! Interface for visiting geometry volumes
+class GeantVolumeAccessor final
+    : public VolumeAccessorInterface<G4LogicalVolume const*,
+                                     G4VPhysicalVolume const*>
+{
+  public:
+    //! Outgoing volume node from an instance
+    VolumeRef volume(VolumeInstanceRef parent) final
+    {
+        CELER_EXPECT(parent);
+        auto result = parent->GetLogicalVolume();
+        CELER_ENSURE(result);
+        return result;
+    }
+
+    //! Outgoing instance edges from a volume
+    ContainerVolInstRef children(VolumeRef parent) final
+    {
+        CELER_EXPECT(parent);
+        ContainerVolInstRef result;
+        for (auto i : range(parent->GetNoDaughters()))
+        {
+            result.push_back(parent->GetDaughter(i));
+        }
+        return result;
+    }
+};
+}  // namespace
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct IDs from world volume.
+ */
+GeantVolumeInstanceMapper::GeantVolumeInstanceMapper(G4PV const& world)
+{
+    // Track volumes that aren't normal/replica/param
+    std::vector<std::string> invalid_volumes;
+
+    // Visit (depth-first) volume instances inside the world
+    VolumeVisitor visit_vol{GeantVolumeAccessor{}};
+    visit_vol(
+        [&](G4VPhysicalVolume const* pv, int) {
```

Comment:
```suggestion
        [&invalid_volumes](G4VPhysicalVolume const* pv, int) {
```
makes it easier to distinguish internal and external variable for the lambda.

Reply:
Also requires capturing `this`